relay-server/src/actors/project_upstream.rs

```diff
@@ -47,11 +48,17 @@ pub struct GetProjectStates {
     pub no_cache: bool,
 }
 
+/// The response of the projects states requests.
+///
+/// A [`ProjectKey`] is either pending or has a result, it can not appear in both and doing
+/// so is undefined.
 #[derive(Debug, Deserialize, Serialize)]
 #[serde(rename_all = "camelCase")]
 pub struct GetProjectStatesResponse {
     #[serde(default)]
     pub configs: HashMap<ProjectKey, ErrorBoundary<Option<ProjectState>>>,
+    #[serde(default)]
+    pub pending: Vec<ProjectKey>,
```

Comment:
We could also define precedence of when both are present, for instance that the supplied config always wins over the pending state.

The initial approach was to do exactly what you are suggesting. Making serialization work in an unobtrusive way (supporting both v2 and v3 without much overhead) is slightly harder in that scenario, but still possible.

Reply:
what is the advantage of the currently implemented schema? the protocol is not backwards-compatible anyway, even if it is at the syntactical level, we cannot send v3 to a v2 relay. if we accidentally send v3 response to v2 relay, deserialization failure (unknown value "pending") is actually safer IMO than assuming the project does not exist. The latter causes silent data dropping while the former throws errors everywhere. (Presumably we will see alerts either way, but customer relays may have less monitoring)

also, will we write pending state into redis, and if so, how?

Reply:
This implementation was just massively easier to implement and read than the other one after trying both out.

In theory, we can send this down to a V2 Relay, and this is also how the endpoint is implemented. Such a Relay would assume the now missing project config as missing and stop polling. Since that would be incorrect in practice, the new key will never be sent to a V2 Relay based on the endpoint implementation.